### DISCUSSION TITLE
Modify data in FutureProvider

### DISCUSSION THREAD

**Discussion by buraktabn** [rrousselGit/riverpod#734] [Ideas]
I could just use `StateNotifier` but if I could edit the data in `FutureProvider` like `StateProvider`, that would clear some boilerplate and it would make my life easy at least.

#### Use Case
Let's say we have a `FutureProvider` that fetches comments of a post. When a user adds a new comment to the post, that'd be easier to add the new comment to the list in `FutureProvider`. 

RiverPod could have something like `FutureStateProvider`  that combines both `FutureProvider` and `StateProvider`.

```dart
final myProvider = FutureStateProvider<List<Comment>>((ref) => fetchComments());
...
ref.read(myProvider).state = [...ref.read(myProvider).data?.value ?? [], myNewComment];
```

**Comment by andreadaoud** [comment]
+1 on this issue. Another use case would be a settings page for a remote device, where you mainly load settings from remote device, but when the user change settings, you want to "force" some value to be the changed settings, instead of loading them again (loading again introduces delay). I think we should be able to change the "cached" value inside FutureProvider.

@rrousselGit Would you please give some comments on this topic?

**Reply by rrousselGit** [reply]
That's part of the new notifier api.

**Reply by andreadaoud** [reply]
> That's part of the new notifier api.

Thank you very much for your answer! Could you please give a link to the documentation or example on this feature?

**Reply by andreadaoud** [reply]
Did you mean that we should use `StateNotifierProvider`? We definitely can, but this would make it impossible to use `ref.invalidate`, which is sometimes very inconvenient.

**Reply by rrousselGit** [reply]
Not StateNotifierProvider. The new AsyncNotifier instead
It's the fusion between FutureProvider and StateNotifierProvider basically. But you'll see more about it when it's available.

**Reply by roubachof** [reply]
wonderful! I was going to ask the exact same question : )
For now I have created a `FutureStateNotifier` handling all loading state has well as the result.
`AsyncNotifier` looks great :)
